examples/src/main/java/com/mapbox/navigation/examples/core/MapMatchingActivity.kt

```diff
@@ -0,0 +1,396 @@
+        override fun onSessionStateChanged(tripSessionState: TripSessionState) {
+            when (tripSessionState) {
+                STARTED -> stopLocationUpdates()
+                STOPPED -> {
+                    startLocationUpdates()
+                    navigationMapboxMap?.removeRoute()
+                    updateCameraOnNavigationStateChange(false)
+                }
+            }
+        }
+    }
+
+    private fun updateCameraOnNavigationStateChange(navigationStarted: Boolean) {
+        navigationMapboxMap?.apply {
+            if (navigationStarted) {
+                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_GPS)
+                updateLocationLayerRenderMode(RenderMode.GPS)
+            } else {
+                updateCameraTrackingMode(NavigationCamera.NAVIGATION_TRACKING_MODE_NONE)
+                updateLocationLayerRenderMode(RenderMode.COMPASS)
+            }
+        }
+    }
+
+    private inner class MyLocationEngineCallback(activity: MapMatchingActivity) :
+        LocationEngineCallback<LocationEngineResult> {
+
+        private val activityRef = WeakReference(activity)
+
+        override fun onSuccess(result: LocationEngineResult) {
+            activityRef.get()?.locationComponent?.forceLocationUpdate(result.lastLocation)
+        }
+
+        override fun onFailure(exception: java.lang.Exception) {
+            Timber.i(exception)
+        }
+    }
+
+    private inner class TestUrlSkuTokenProvider : UrlSkuTokenProvider {
+        override fun obtainUrlWithSkuToken(resourceUrl: String, querySize: Int): String {
+            val urlBuilder = StringBuilder(resourceUrl)
+            when (querySize == 0) {
+                true -> urlBuilder.append("?")
+                false -> urlBuilder.append("&")
+            }
+            urlBuilder.append("sku=${Utils.getMapboxAccessToken(applicationContext)}")
+            return urlBuilder.toString()
+        }
+    }
+
+    private inner class TestRerouteController : RerouteController {
+        override val state = RerouteState.Idle
+
+        override fun reroute(routesCallback: RerouteController.RoutesCallback) {
+            Timber.d("TestRerouteController reroute")
+
+            ifNonNull(
+                locationComponent?.lastKnownLocation,
+                destination
+            ) { location, destination ->
+                fetchRouteWithDirectionsApi(location.toPoint(), destination)
+            }
+        }
+
+        override fun interrupt() {
+            // do nothing
+        }
+
+        override fun registerRerouteStateObserver(rerouteStateObserver: RerouteController.RerouteStateObserver): Boolean {
+            throw Exception("Not yet implemented")
+        }
+
+        override fun unregisterRerouteStateObserver(rerouteStateObserver: RerouteController.RerouteStateObserver): Boolean {
+            throw Exception("Not yet implemented")
+        }
```

Comment:
Looking at these `Not yet implemented` functions wondering if we should extract them out from `RerouteController` into a separate `interface` following the [Interface segregation SOLID principle](https://en.wikipedia.org/wiki/Interface_segregation_principle) 🤔 What do you think? cc @RingerJK

Reply:
yeah, it should work for us 👍

Reply:
#3450

Reply:
Those listeners are essential though. An app can always get the current `RerouteController` reference from `MapboxNavigation` and the reroute state observers should always be functional in my opinion.

Also, we're not fully reimplementing the reroute controller here, we're completely ignoring the state.

Reply:
As an example, the UI SDK might be using the reroute state listener to make UI changes, but the developer might opt to override the reroute controller in the Core SDK. Because of this scenario, the state and the state listeners should always be fully reimplemented, I don't see a room for us to split the interfaces. Unless we can move the state and state listeners handling to `MapboxNavigation` instead of having it as part of the `RerouteController`?

Reply:
I thought smtl like 

```
interface RerouteController : RerouteControllerObserversContract {
...
}
```

where 

```
RerouteControllerObserversContract { 
  fun registerRerouteStateObserver ...
  fun unregisterRerouteStateObserver ...
}
```

Reply:
That would still require the end-user overriding the `RerouteController` to also override the observers so doesn't really benefit the segregation :(

Reply:
yeah, agree. But now SOLID 😄

Reply:
Yeah don't see the benefit of `RerouteControllerObserversContract` either.

Not sure if I completely follow @LukasPaczos but my point was more around the lines that if you as a client decide to implement a custom `RerouteController` you may want to implement how to be notified of state changes differently or even not implement those at all (as in this example) i.e. you shouldn't be forced to implement `registerRerouteStateObserver` and `unregisterRerouteStateObserver`. I think the key parts of `RerouteController` are `state`, `reroute` and `interrupt` the rest should be optional IMO. I'm also wondering if we can improve the `reroute(routesCallback: RoutesCallback)` in https://github.com/mapbox/mapbox-navigation-android/pull/3437#discussion_r471790777

Reply:
We might be internally depending on the reroute controller state though, and we might need to have state observers available. This means that someone overwriting the default logic also has to provide those features. Or as I mentioned above, maybe there's an opportunity to move those state observers directly into the `MapboxNavigation` class and change the interface to accommodate it.

Reply:
> We might be internally depending on the reroute controller state though, and we might need to have state observers available.

Yeah, that's `MapboxRerouteController`.

> This means that someone overwriting the default logic also has to provide those features.

Yeah, that's expected IMO. I mean, if you decide to implement your own `RerouteController` you may want to implement the state machine differently and that "freedom" is actually what `RerouteController` should give you.

> Or as I mentioned above, maybe there's an opportunity to move those state observers directly into the `MapboxNavigation` class and change the interface to accommodate it.

Yeah, that'd align well with the rest of the observers as they're all exposed through `MapboxNavigation`. So that'd be consistent but with the trade-off of not being able to implement state listeners logic differently. Which yeah maybe in favor of consistency is a minor trade-off that we can skip 🚀

Reply:
@LukasPaczos @RingerJK Could either of you follow up with a draft PR when you have a chance?

BTW to unblock this PR let's move the discussions to https://github.com/mapbox/mapbox-navigation-android/issues/3450